File: icp/examples/event_plane/publisher.py
# SPDX-FileCopyrightText: Copyright (c) 2024-2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


import argparse
import asyncio
import uuid

from triton_distributed.icp import NatsEventPlane, Topic


async def main(component_id, event_type, publisher_id, event_count):
    server_url = "nats://localhost:4222"
    event_plane = NatsEventPlane(server_url, component_id)

    await event_plane.connect()

    try:
        topic = Topic(["publisher", str(publisher_id)])

        for i in range(event_count):
            payload = f"Payload from publisher {publisher_id}".encode()
            event = event_plane.create_event(event_type, topic, payload)
            await event_plane.publish(event)
            print(f"Published event from publisher {publisher_id}: {event.event_id}")
            await asyncio.sleep(0.01)
    finally:
        await event_plane.disconnect()


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Event publisher script")
    parser.add_argument(
        "--component_id",
        type=uuid.UUID,
        default=uuid.uuid4(),
        help="Component ID (UUID)",
    )
    parser.add_argument(
        "--event_type", type=str, default="test_event", help="Event type"
    )
    parser.add_argument("--publisher_id", type=int, required=True, help="Publisher ID")
    parser.add_argument(
        "--event-count", type=int, default=10, help="Event count to be published."
    )

    args = parser.parse_args()
    asyncio.run(
        main(args.component_id, args.event_type, args.publisher_id, args.event_count)
    )

File: runtime/rust/src/kv_router/publisher.rs
// SPDX-FileCopyrightText: Copyright (c) 2024-2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use crate::kv_router::{indexer::RouterEvent, protocols::KvCacheEvent, KV_EVENT_SUBJECT};
use crate::{component::Component, DistributedRuntime, Result};
use tokio::sync::mpsc;
use uuid::Uuid;
use tracing as log;
use tracing_subscriber::FmtSubscriber;
use std::fs::OpenOptions;
use std::io::Write;

pub struct KvPublisher {
    tx: mpsc::UnboundedSender<KvCacheEvent>,
}

impl KvPublisher {
    pub fn new(drt: DistributedRuntime, backend: Component, worker_id: Uuid) -> Result<Self> {

        let subscriber = FmtSubscriber::builder()
            // all spans/events with a level higher than TRACE (e.g, debug, info, warn, etc.)
            // will be written to stdout.
            .with_env_filter(tracing_subscriber::EnvFilter::from_default_env())
            // completes the builder.
            .finish();

        tracing::subscriber::set_global_default(subscriber)
        .expect("setting default subscriber failed");
        
        log::info!("Logging Started in Publisher");

        let (tx, rx) = mpsc::unbounded_channel::<KvCacheEvent>();
        let p: KvPublisher = KvPublisher { tx };

        start_publish_task(drt, backend, worker_id, rx);
        Ok(p)
    }

    pub fn publish(&self, event: KvCacheEvent) -> Result<(), mpsc::error::SendError<KvCacheEvent>> {
        log::info!("Attempting to publish event: {:?}", event);
        self.tx.send(event)
    }
}

fn start_publish_task(
    drt: DistributedRuntime,
    backend: Component,
    worker_id: Uuid,
    mut rx: mpsc::UnboundedReceiver<KvCacheEvent>,
) {
    let client = drt.nats_client().client().clone();
    // [FIXME] service name is for metrics polling?
    // let service_name = backend.service_name();
    let kv_subject: String = backend.event_subject(KV_EVENT_SUBJECT);
    log::info!("Publishing to subject: {}", kv_subject);
    _ = drt.runtime().secondary().spawn(async move {
        while let Some(event) = rx.recv().await {
            let router_event: RouterEvent = RouterEvent::new(worker_id, event);
            let data: String = serde_json::to_string(&router_event).unwrap();
            client
                .publish(kv_subject.to_string(), data.into())
                .await
                .unwrap();
        }
    });
}
